include/mra/misc/convolutiondata.h

```diff
@@ -0,0 +1,156 @@
+#ifndef MRA_CONVOLUTIONDATA_H
+#define MRA_CONVOLUTIONDATA_H
+
+#include "mra/ops/inner.h"
+#include "mra/misc/gl.h"
+#include "mra/misc/hash.h"
+#include "mra/misc/misc.h"
+#include "mra/misc/types.h"
+#include "mra/misc/adquad.h"
+#include "mra/misc/twoscale.h"
+#include "mra/misc/platform.h"
+#include "mra/misc/autocorr.h"
+#include "mra/tensor/tensorview.h"
+
+#define MRA_MAX_LX 7
+namespace mra {
+
+  // template <typename T>
+  // struct ConvolutionData {
+  //   Tensor<T, 2> R, T;
+
+  // }
+
+  template <typename T, Dimension NDIM>
+  class Convolution {
+
+    private:
+      size_type K;
+      Level n;
+      int npt;                                      // number of quadrature points
+      T expnt;
+      T coeff;
+      Translation lx;
+      const T* quad_x;                              // quadrature points
+      const T* quad_w;                              // quadrature weights
+      Tensor<T, 3> autocorrcoef;                    // autocorrelation coefficients
+      Tensor<T, 3> c;                               // autocorrelation coefficients
+      std::map<Key<NDIM>, Tensor<T, 2>> rnlijcache; // map for storing rnlij matrices
+      std::map<Key<NDIM>, Tensor<T, 1>> rnlpcache; // map for storing rnlp matrices
+      std::mutex cachemutex;                        // mutex for thread safety
+
+
+      void autoc(){
+        auto c_view = c.current_view();
+        auto autocorr_view = autocorrcoef.current_view();
+        detail::autocorr_get<T>(K, autocorr_view);
+        c_view = 0.0;
+        std::array<Slice,NDIM> slices = {Slice(0, K), Slice(0, K), Slice(0, 2*K)};
+        c_view(slices) = autocorr_view(slices);
+        slices = {Slice(0, K), Slice(0, K), Slice(2*K, 4*K)};
+        c_view(slices) = autocorr_view(slices);
+      }
+
+      // projection of a Gaussian onto double order polynomials
+      Tensor<T, 1> make_rnlp(const Level n, Translation lx) {
+        mra::Key<NDIM> key(n, std::array<Translation, NDIM>({lx}));
+        auto it = rnlpcache.find(key);
+        if (it != rnlpcache.end()) {
+          return std::move(it->second);
+        }
+        Tensor<T, 1> rnlp(2*K);
+        auto rnlp_view = rnlp.current_view();
+        rnlp_view = 0.0;
+
+        if (lx < 0) lx = -lx-1;
+        T scaledcoeff  = coeff*std::pow(0.5, 0.5*n);
+        T beta = expnt * std::pow(T(0.25), T(n));
+        T h = 1.0/std::sqrt(beta);
+        T nbox = 1.0/h;
+        if (nbox < 1) nbox = 1;
+        h = 1.0/nbox;
+
+        T sch = std::abs(scaledcoeff*h);
+
+        T argmax = std::abs(std::log(1e-22/sch));
+
+        for (size_type box=0; box<nbox; ++box){
+          T xlo = box*h + lx;
+          if (beta*xlo*xlo > argmax) break;
+
+          for (size_type i=0; i<npt; ++i){
+            T* phix = new T[2*K];
```

Comment:
This is never released.

Reply:
Sorry I do not follow this comment. What never gets released?

Reply:
You're allocating memory but never release it.

Reply:
Addressed in recent commit.